### DISCUSSION TITLE
inject settings.xml before scan

### DISCUSSION THREAD

**Discussion by ramarro123** [renovatebot/renovate#24491] [Request Help]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, select which platform you are using.

None

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

hi,

i have a set of git project, that (will) be scanned by a single renovate job inserted into a separata pipeline with gitlab scheduled run (i basically followed a guide on internet hehehe)

i have a question, during build, the current build pipeline copy a file "settings.xml" into .m2/settings.xml before running mvn clean install.

so, no .m2 file are committed on repository, this because the repository is agnostic, and it could change based on the fact that the command is run inside company lan (we have a private nexus) or over intenret (mavencentral)
for compliance, during the release build we have to use the private nexus.

now, back to my question, if i store the "settings.xml" in my renovate_scheduler project, it's possible to "copy" or "assign" it to every scan it will perform (based on config.js)? i did try setting   MAVEN_ARGS: "-s  blablal" but it didn't worked unfortunately :) 

### What you tried so far.

set the mvn arg from export or pipeline variable

### Relevant debug logs

<details><summary>Logs</summary>

```
Copy/paste the relevant log(s) here, between the starting and ending backticks
```

</details>

**Comment by rarkins** [comment]
Why? What are you trying to achieve?

**Reply by ramarro123** [reply]
as said, due to company policy, i can't have every project commit the .m2 folder on the gitlab project.

i need to inject the settings.xml before it actually run the mvn <stuff>

access to nexus is limited with a token/password that every developer must config on his own desktop, having a committed settings.xml just didn't fit our scenario.

do you think there is a way to cover this scenario? a "preScanTask" hook?

**Reply by rarkins** [reply]
Configure hostRules instead

Read https://docs.renovatebot.com/getting-started/private-packages/

**Reply by ramarro123** [reply]
host rule won't work, as i have to "rewrite" the default maven behaviour.

consider this settings.xml (added at the end of comment)
we are using a "mirrorOf" central to make sure that all our project(s) get redirect to a verified repository of the bank, that "should" be validated and contains an allowed "subset" of central.

In this scenario hostRule won't work (or not unless i am not able to inject the mirror config for mavencentral)

```



<settings xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
          xmlns="http://maven.apache.org/SETTINGS/1.0.0"
          xsi:schemaLocation="http://maven.apache.org/SETTINGS/1.0.0
                      http://maven.apache.org/xsd/settings-1.0.0.xsd">

    <localRepository>/opt/mvn/.m2</localRepository>

    <mirrors>
        <mirror>
            <id>Nexus</id>
            <url>https://artifactrepository.bansel.it/nexus/content/groups/public</url>
            <mirrorOf>central</mirrorOf>
        </mirror>
    </mirrors>

    <profiles>
        <profile>
            <id>gitlab-maven</id>
            <repositories>

                <repository>
                    <releases>
                        <enabled>true</enabled>
                    </releases>
                    <snapshots>
                        <enabled>true</enabled>
                    </snapshots>
                    <id>gitlab-maven</id>
                    <name>Gitlab Maven Core</name>
                    <url>${env.CI_API_V4_URL}/groups/155/-/packages/maven</url>
                </repository>

            </repositories>
            <pluginRepositories>

                <pluginRepository>
                    <id>gitlab-maven</id>
                    <url>${env.CI_API_V4_URL}/groups/155/-/packages/maven</url>
                </pluginRepository>

            </pluginRepositories>
            <properties>
                <altDeploymentRepository>
                    gitlab-maven::default::${env.CI_API_V4_URL}/projects/${env.CI_PROJECT_ID}/packages/maven
                </altDeploymentRepository>
            </properties>
        </profile>
    </profiles>

    <activeProfiles>
        <activeProfile>gitlab-maven</activeProfile>
    </activeProfiles>

    <servers>
        <server>
            <id>gitlab-maven</id>
            <configuration>
                <httpHeaders>
                    <property>
                        <name>Deploy-Token</name>
                        <value>${env.CI_DEPLOY_PASSWORD}</value>
                    </property>
                </httpHeaders>
            </configuration>
        </server>
    </servers>
</settings>
```

**Reply by rarkins** [reply]
Then use packageRules with registryUrls too

I asked before why, you didn't really give enough details, so now I'm guessing. You haven't even said which part _fails_ without this

**Reply by ramarro123** [reply]
ok, sorry for not giving enough detail, well if i won't use this settings.xml it fail to look-up for package that are hosted on this private nexus.

i will look at packageRules thank you for the support

**Reply by rarkins** [reply]
You can use `packageRules` with `registryUrls` to override detected registries.
